sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/repository/metric/InMemoryMetricsRepository.java

```diff
@@ -46,46 +47,68 @@
      */
     private Map<String, Map<String, LinkedHashMap<Long, MetricEntity>>> allMetrics = new ConcurrentHashMap<>();
 
-    private final ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
 
+    private Map<String, ReadWriteLock> appLockMap = new ConcurrentHashMap(); //lock seperate by app
 
-    @Override
-    public void save(MetricEntity entity) {
-        if (entity == null || StringUtil.isBlank(entity.getApp())) {
-            return;
-        }
-        readWriteLock.writeLock().lock();
-        try {
-            allMetrics.computeIfAbsent(entity.getApp(), e -> new HashMap<>(16))
-                    .computeIfAbsent(entity.getResource(), e -> new LinkedHashMap<Long, MetricEntity>() {
-                        @Override
-                        protected boolean removeEldestEntry(Entry<Long, MetricEntity> eldest) {
-                            // Metric older than {@link #MAX_METRIC_LIVE_TIME_MS} will be removed.
-                            return eldest.getKey() < TimeUtil.currentTimeMillis() - MAX_METRIC_LIVE_TIME_MS;
-                        }
-                    }).put(entity.getTimestamp().getTime(), entity);
-        } finally {
-            readWriteLock.writeLock().unlock();
-        }
 
+    @Override
+    public void save(MetricEntity entity, String app) {
+        allMetrics.computeIfAbsent(app, e -> new HashMap<>(16)).computeIfAbsent(entity.getResource(), e -> new LinkedHashMap<Long, MetricEntity>(100) {
+            @Override
+            protected boolean removeEldestEntry(Map.Entry<Long, MetricEntity> eldest) {
+                return eldest.getKey() < System.currentTimeMillis() - MAX_METRIC_LIVE_TIME_MS;
+            }
+        }).put(entity.getTimestamp().getTime(), entity);
     }
 
     @Override
-    public void saveAll(Iterable<MetricEntity> metrics) {
+    public void saveAll(Iterable<MetricEntity> metrics, String app) {
+
         if (metrics == null) {
             return;
         }
-        readWriteLock.writeLock().lock();
+
+        if (appLockMap.get(app) == null) { //DCL to the same app concurrent request
+
+            synchronized (this) {
+                if (appLockMap.get(app) == null) {
+                    appLockMap.put(app, new ReentrantReadWriteLock());
+                    allMetrics.put(app, new HashMap<>(256));
+                    lockAndWriteData(appLockMap.get(app), app, metrics);
+                } else { //
```

Comment:
why synchronized `InMemoryMetricsRepository`  here ?  it seems synchronized  `appLockMap` will be better. And                     `allMetrics.put(app, new HashMap<>(256));` do duplicate operation which also do in `save` method

Reply:
1.first question. synchronize is avoid to repeat init the app-map。 it only affect when the request is first arrival
2.about duplicate operation. only to passed the unit test. have no actually meanings.

Reply:
the synchonized  is only affect the app init.for example. we have 40 app.   the synchorinze only affect in 40 app firest report data.after init. after 40 locks generate, all code where arrival else.

Reply:
init a lock blocked all the  other invoker’s call is not a highly performance design.  Another optional design is defined another lock  which control the `app lock` init not the `InMemoryMetricsRepository` itself.  And we should keep the code clean and simple, don't  repeat your self.

Reply:
1.about duplicate operation.i will remove the duplication code in save method and update the unit test.
2.about the synchronize in init.   it  only  invoked in init period.  i think is could accepted. maybe i should init lock  in the app register ?